File: swift/collatz-conjecture/Sources/CollatzConjecture/CollatzConjecture.swift
class CollatzConjecture {
  struct InvalidNumber: Error {}

  static func steps(_ number: Int) throws -> Int {
    guard number > 0 else {
      throw InvalidNumber()
    }

    if number == 1 {
      return 0
    }

    let nextNumber = number.isMultiple(of: 2) ? number / 2 : 3 * number + 1
    return try 1 + steps(nextNumber)
  }
}

File: c/collatz-conjecture/collatz_conjecture.c
#include "collatz_conjecture.h"

int steps(int start)
{
    if (start < 1)
        return ERROR_VALUE;

    if (start == 1)
        return 0;

    if (start % 2 == 0)
        return 1 + steps(start / 2);

    return 1 + steps(start * 3 + 1);
}
